src/main/java/org/springframework/integration/kafka/listener/KafkaTopicOffsetManager.java

```diff
@@ -0,0 +1,505 @@
+import org.springframework.integration.kafka.core.KafkaMessage;
+import org.springframework.integration.kafka.core.KafkaMessageBatch;
+import org.springframework.integration.kafka.core.KafkaTemplate;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.Result;
+import org.springframework.integration.kafka.core.TopicNotFoundException;
+import org.springframework.integration.kafka.support.ProducerFactoryBean;
+import org.springframework.integration.kafka.support.ProducerMetadata;
+import org.springframework.integration.kafka.support.ZookeeperConnect;
+import org.springframework.integration.kafka.util.LoggingUtils;
+import org.springframework.integration.kafka.util.MessageUtils;
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.RetryContext;
+import org.springframework.retry.backoff.ExponentialBackOffPolicy;
+import org.springframework.retry.policy.SimpleRetryPolicy;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+
+/**
+ * Implementation of an {@link OffsetManager} that uses a Kafka topic as the underlying support.
+ * For its proper functioning, the Kafka server(s) must set {@code log.cleaner.enable=true}. It relies on the property
+ * {@code cleanup.policy=compact} to be set on the target topic, and if the topic is not found,
+ * it will create a topic with the appropriate settings.
+ *
+ * @author Marius Bogoevici
+ */
+@SuppressWarnings({"unchecked", "rawtypes"})
+public class KafkaTopicOffsetManager extends AbstractOffsetManager implements InitializingBean {
+
+	private static final KeyEncoderDecoder KEY_CODEC = new KeyEncoderDecoder();
+
+	private static final LongEncoderDecoder VALUE_CODEC = new LongEncoderDecoder();
+
+	public static final String CLEANUP_POLICY = "cleanup.policy";
+
+	public static final String CLEANUP_POLICY_COMPACT = "compact";
+
+	public static final String DELETE_RETENTION = "delete.retention.ms";
+
+	public static final String SEGMENT_BYTES = "segment.bytes";
+
+	private final ZookeeperConnect zookeeperConnect;
+
+	private final ConnectionFactory connectionFactory;
+
+	private final String topic;
+
+	private final KafkaTemplate kafkaTemplate;
+
+	private final ConcurrentMap<Partition, Long> data = new ConcurrentHashMap<Partition, Long>();
+
+	private String compressionCodec = "default";
+
+	private Producer<Key, Long> producer;
+
+	private int maxSize = 10 * 1024;
+
+	private int maxQueueBufferingTime = 1000;
+
+	private int segmentSize = 25 * 1024;
+
+	private int retentionTime = 60000;
+
+	private int replicationFactor;
+
+	private int maxBatchSize = 200;
+
+	private boolean batchWrites = true;
+
+	private int requiredAcks = 1;
+
+	public KafkaTopicOffsetManager(ZookeeperConnect zookeeperConnect,
+			ConnectionFactory connectionFactory, String topic, Map<Partition, Long> initialOffsets) {
```

Comment:
Also we can close connectionFactory after reading offset data, and save a socket for future ;).